### PR TITLE
[ViP-DeepLab] Add wSTQ in numpy for PVPS dataset.

### PR DESCRIPTION
**Pull request by meijieru**
This PR adds wSTQ implementation in numpy.
It also adds the unit test to guarantee the compatibility of tf & numpy implementation.
Fix a dtype error for tf impl.

### CODE REVIEW

File: evaluation/numpy/segmentation_and_tracking_quality.py [hunk 1]
```diff
@@ -108,7 +126,27 @@ def get_semantic(self, y: np.ndarray) -> np.ndarray:
     """Returns the semantic class from a panoptic label map."""
     return y >> self._label_bit_shift
 
-  def update_state(self, y_true: np.ndarray, y_pred: np.ndarray, sequence_id=0):
+  def _get_or_update_confusion_matrix(self,
+                                      y_true: np.ndarray,
+                                      y_pred: np.ndarray,
+                                      weights: Optional[np.ndarray] = None,
+                                      cm: Optional[np.ndarray] = None):
```
Comment (aquariusjay) [inline evaluation/numpy/segmentation_and_tracking_quality.py:133]: nit: rename cm as confusion_matrix

File: evaluation/numpy/segmentation_and_tracking_quality.py [hunk 2]
```diff
@@ -136,25 +177,19 @@ def update_state(self, y_true: np.ndarray, y_pred: np.ndarray, sequence_id=0):
                                 semantic_label, self._num_classes)
       semantic_prediction = np.where(semantic_prediction != self._ignore_label,
                                      semantic_prediction, self._num_classes)
+    # TODO(meijieru): weight the confusion matrix.
```
Comment (aquariusjay) [inline evaluation/numpy/segmentation_and_tracking_quality.py:180]: Will we resolve this TODO or this is more like a note?

File: evaluation/numpy/segmentation_and_tracking_quality.py [hunk 3]
```diff
@@ -136,25 +177,19 @@ def update_state(self, y_true: np.ndarray, y_pred: np.ndarray, sequence_id=0):
...
     if sequence_id in self._iou_confusion_matrix_per_sequence:
-      idxs = (np.reshape(semantic_label, [-1]) <<
```
Comment (aquariusjay) [inline evaluation/numpy/segmentation_and_tracking_quality.py:140]: The original implementation contains shift operations, which significantly speeds up the computation for large datasets.
Is it possible to keep it?

**Review by aquariusjay — COMMENTED** [review]
Thanks, meijieru!
A few suggestions. PTAL.

File: evaluation/numpy/segmentation_and_tracking_quality.py (hunk 3 above)
Comment (meijieru) [reply evaluation/numpy/segmentation_and_tracking_quality.py:140]: If we could make the assumption that the only possible weights are in `[0.5, 1]`, then we could also encode the weights when doing the `np.unique`.

**Comment by markweberdev** [comment]
Hi @meijieru,

Thanks for your work, I appreciate that! I ran some quick performance test to see the difference when using the PR's code. On my local machine, I was able to observe a runtime increase of a bit more than 70%. I would address this mainly due to how the confusion matrix is computed. 

Given that, we run this code in some time critical environments (like benchmark server), I would suggest to not alter the code directly, but make a weighted STQ subclass that inherits from STQ. This way, we could have both performance for the base case, but also support for the weighted STQ. 

What do you think @aquariusjay ?

File: evaluation/numpy/segmentation_and_tracking_quality.py (hunk 3 above)
Comment (aquariusjay) [reply evaluation/numpy/segmentation_and_tracking_quality.py:140]: We could raise error if weights are not in [0.5, 1] and add comment that this assumption is need for this part of code. What do you think?

**Comment by aquariusjay** [comment]
@markweberdev Thanks for reviewing the code. The suggestion sounds good.

@meijieru Could you please try to update the code to maintain the original computation speed? It would be great if wSTQ could be also efficiently computed. What do you think?

**Comment by meijieru** [comment]
Several updates

1. We maintain the original computation speed, and remove sparse addition in the original implementation.
2. Fix a problem of precision error in tf wSTQ.

Would you please check again? Thx.

**Comment by aquariusjay** [comment]
Looks great on my end. Thanks for updating the code, @meijieru 

Please wait for the input from @markweberdev

File: evaluation/numpy/segmentation_and_tracking_quality.py [hunk 4]
```diff
@@ -122,9 +176,12 @@ def update_state(self, y_true: np.ndarray, y_pred: np.ndarray, sequence_id=0):
       sequence_id: The optional ID of the sequence the frames belong to. When no
         sequence is given, all frames are considered to belong to the same
         sequence (default: 0).
+      weights: The weights for each pixel with the same shape of `y_true`.
```
Comment (markweberdev) [inline evaluation/numpy/segmentation_and_tracking_quality.py:179]: Minor: `The optional weights ... Defaults to using the same weight for every pixel.`

**Review by markweberdev — COMMENTED** [review]
Thanks for speeding this up!

Besides a minor comment, I have only one suggestion. I would like to keep the numpy test file self-contained, so users can just copy the numpy folder, work with it without any tf or other deeplab2 dependencies. Thus, I'd suggest to take the compatibility test, which I like, and move it to the tensorflow test file of the metric. What do you think?

**Review by Billy8927 — APPROVED** [review]

**Comment by meijieru** [comment]
Updated, thanks for the suggestions.

**Review by markweberdev — APPROVED** [review]
Looks good to me!

**Pull request merged**